File: statcheck-ml/r/extract.R
# Read a PDF the way the R port must, and hand back text the rest can use.
#
# The R port uses pdftools, which bundles poppler. R has no other maintained
# PDF reader, so this is a constraint and not a choice. Measured on 198 holdout
# documents with 323 gold results, pdftools holds 0.916 of them and keeps 0.912
# after the prefilter, against 0.929 for the Python port.
#
# pdf_text() is not the same as the pdftotext command line. It reads more. Never
# measure the R port by borrowing a pdftotext number.
#
# pdftools hangs on some files, twice in the 198 measured, where the command
# line reads the same files without trouble. Each document therefore gets a time
# limit, and a document that passes it returns empty text so the caller can
# continue with the rest of the corpus.
#
# R.utils supplies the time limit. It belongs in Suggests, not Imports: without
# it the reader still works and only loses the limit.

# `normalize.R` must be sourced first. Finding it from inside a sourced file is
# unreliable in R, and guessing wrongly is worse than asking the caller.
if (!exists("normalize_text", mode = "function")) {
  stop("source normalize.R before extract.R", call. = FALSE)
}

#' Is the time limit available?
has_timeout <- function() {
  requireNamespace("R.utils", quietly = TRUE)
}

#' Read one PDF and return normalised text.
#'
#' @param path The PDF.
#' @param spec The parsed normalisation spec.
#' @param timeout Seconds to allow one document. Ignored when R.utils is absent.
#' @return Normalised text, or "" when the document could not be read.
read_pdf_text <- function(path, spec, timeout = 60) {
  read_it <- function() paste(pdftools::pdf_text(path), collapse = "")

  txt <- tryCatch({
    if (has_timeout()) {
      R.utils::withTimeout(read_it(), timeout = timeout, onTimeout = "silent")
    } else {
      # Without R.utils there is no limit, and a bad document can stop the run.
      # Warn once so the cause is visible rather than mysterious.
      if (!isTRUE(getOption("statcheckml.timeout.warned"))) {
        warning("R.utils is not installed, so no time limit applies to ",
                "pdftools. Install R.utils to stop one document halting a run.",
                call. = FALSE)
        options(statcheckml.timeout.warned = TRUE)
      }
      read_it()
    }
  }, error = function(e) "")

  if (is.null(txt) || length(txt) == 0 || is.na(txt[1])) return("")
  normalize_text(txt, spec)
}

File: statcheck-ml/js/extract.js
// Read a PDF in the browser, and hand back text the rest of the system can use.
//
// The browser port uses PDF.js. It is the engine a browser can use without
// downloading a second binary, it is under Apache-2.0, and it reads almost as
// much as the engine the models were trained on. Measured on 198 holdout
// documents with 323 gold results, PDF.js holds 0.926 of them against 0.929 for
// PyMuPDF, and keeps 0.923 after the prefilter against 0.929.
//
// PDF.js reports text as pieces with a position and no lines at all. Where a
// line ends is therefore a decision this file makes, and every port must make
// the same one, or each port reads a different document. `joinItems` holds that
// decision and `spec/normalize.json` holds everything after it.
//
// Usage:
//   import { extractText } from './extract.js';
//   const text = await extractText(pdfjs, arrayBuffer, spec);

import { createNormalizer } from './normalize.js';

/** The vertical distance, in PDF units, that starts a new line. */
const NEW_LINE_AT = 2;

/**
 * Join the pieces PDF.js returns into lines.
 *
 * @param {object} textContent The value of `page.getTextContent()`.
 * @returns {string}
 */
export function joinItems(textContent) {
  const parts = [];
  let lastY = null;
  for (const item of textContent.items) {
    if (item.str === undefined) continue;
    const y = item.transform ? Math.round(item.transform[5]) : null;
    if (lastY !== null && y !== null && Math.abs(y - lastY) > NEW_LINE_AT) {
      parts.push('\n');
    }
    parts.push(item.str);
    if (item.hasEOL) parts.push('\n');
    lastY = y;
  }
  return parts.join('');
}

/**
 * Read every page of a PDF and return normalised text.
 *
 * The normalisation stage runs here, so the caller receives text that does not
 * depend on which engine read the file.
 *
 * @param {object} pdfjs The `pdfjs-dist` module.
 * @param {ArrayBuffer|Uint8Array} data The PDF itself.
 * @param {object} spec Parsed `spec/normalize.json`.
 * @param {(page: number, total: number) => void} [onProgress]
 * @returns {Promise<{text: string, pages: number, info: object}>}
 */
export async function extractText(pdfjs, data, spec, onProgress) {
  const { normalize } = createNormalizer(spec);
  const bytes = data instanceof Uint8Array ? data : new Uint8Array(data);
  const doc = await pdfjs.getDocument({ data: bytes, verbosity: 0 }).promise;

  let raw = '';
  try {
    for (let i = 1; i <= doc.numPages; i++) {
      const page = await doc.getPage(i);
      raw += joinItems(await page.getTextContent()) + '\n';
      page.cleanup();
      if (onProgress) onProgress(i, doc.numPages);
    }
  } finally {
    // Releasing the document is optional, and its name has changed between
    // versions. A failure here must never discard the text already read.
    try {
      if (typeof doc.destroy === 'function') await doc.destroy();
      else if (typeof doc.cleanup === 'function') await doc.cleanup();
    } catch { /* nothing to release */ }
  }

  const { text, info } = normalize(raw);
  return { text, pages: doc.numPages, info };
}
